### PR TITLE
Added disable signup option

### PR DESCRIPTION
**Pull request by Skiepp**
Signed-off-by: skiepp93 <skiepp93@hotmail.it>

Resolves https://github.com/artifacthub/hub/issues/2731

Option:

```yaml
server:
  allowUserSignUp: false
```
~~Not reuired -- default to true~~ ->After @tegioz suggestion, changed to required

### CODE REVIEW

**Review by tegioz — CHANGES_REQUESTED** [review]
Hi @Skiepp 👋 

Thanks for the PR!

I've left some comments. @cynthia-sg will follow up with some feedback for the frontend side 🙂

**Comment by Skiepp** [comment]
Thank you @tegioz for the review, I integrated your suggestions 👍

**Comment by tegioz** [comment]
Thanks @Skiepp! 👍

File: web/src/layout/navigation/MobileSettings.tsx [hunk 1]
```diff
@@ -184,19 +187,25 @@ const MobileSettings = (props: Props) => {
...
-                        props.setOpenSignUp(true);
-                      }}
-                      aria-label="Open sign up modal"
-                    >
-                      <div className="d-flex align-items-center">
-                        <FaEdit className="me-2" />
-                        <div>Sign up</div>
-                      </div>
-                    </button>
+                    {allowUserSignUp ? (
```
Comment (cynthia-sg) [inline web/src/layout/navigation/MobileSettings.tsx:190]: Can you use **&&** operator? https://github.com/artifacthub/hub/blob/8cb2f3824354ea3dec3c5556e9164bf0700e5249/web/src/layout/navigation/Navbar.tsx#L67

**Review by cynthia-sg — CHANGES_REQUESTED** [review]
Thanks for your PR! 

Could you please take a look to the frontend tests? You can use the command `yarn test` into `/web` to run all frontend tests. Right now, tests for `Navbar` and `MobileSettings` components are failing. You are going to need adding this block before each test (`beforeEach`):
```
Object.defineProperty(document, 'querySelector', {
  value: (selector: any) => {
    switch (selector) {
      case `meta[name='artifacthub:allowUserSignUp']`:
        return {
          getAttribute: () => 'true',
        };
      default:
        return false;
    }
  },
  writable: true,
});
```

**Comment by Skiepp** [comment]
Thank you @cynthia-sg for the review, I integrated your suggestions 👍

After running the tests I got this error, which seems not be related to this PR:
```
Summary of all failing tests
 FAIL  src/api/index.test.tsx
  ● API › API › getValuesSchema › success

    SyntaxError: Expected double-quoted property name in JSON at position 16026
        at JSON.parse (<anonymous>)

      37 |
      38 | const getData = (fixtureId: string): any => {
    > 39 |   return require(`./__fixtures__/index/${fixtureId}.json`) as any;
         |          ^
      40 | };
      41 |
      42 | const getCSRFTokenMock = jest.fn();

      at Runtime._loadModule (node_modules/jest-runtime/build/index.js:1178:39)
      at getData (src/api/index.test.tsx:39:10)
      at Object.<anonymous> (src/api/index.test.tsx:1818:42)
```

File: web/src/api/__fixtures__/index/42.json [hunk 2]
```diff
@@ -412,7 +413,12 @@
               "type": "boolean",
               "title": "Allow adding private repositories to the Hub",
               "default": false
-            }
+            },
+            "allowUserSignUp": {
+              "type": "boolean",
+              "title": "Allow new Users to Sign Up",
+              "default": true
+            },
```
Comment (cynthia-sg) [inline web/src/api/__fixtures__/index/42.json:421]: I think that the test is failing due to the trailing comma.

**Review by cynthia-sg — CHANGES_REQUESTED** [review]
Thanks for the changes.

File: web/src/api/__fixtures__/index/42.json (hunk 2 above)
Comment (Skiepp) [reply web/src/api/__fixtures__/index/42.json]: Whoops, guess I've been spoiled by Dart :)

All checks are working fine now 👍

**Review by cynthia-sg — APPROVED** [review]
LGTM 👍

**Review by tegioz — APPROVED** [review]
👍

**Comment by Skiepp** [comment]
I think right now some backend test are failing due to a return code 403 (Forbidden) instead of an expected code (eg 201 or 400)

This is because if you don't specify anything for the settings "allowUserSignUp", it is implicitly evaluated as false, so the backend will not let you register

You sure it's not better to set allowUserSignUp to true as default behaviour and mark the setting as optional to not break compatibility?

**Comment by tegioz** [comment]
> You sure it's not better to set allowUserSignUp to true as default behaviour and mark the setting as optional to not break compatibility?

The default is still `true` as you defined in the chart default values, it's just that it isn't set in the backend (we usually don't do that for any other default value) 🙂

I would suggest setting the new configuration value to `true` for the users handlers tests below [this line](https://github.com/artifacthub/hub/blob/3866c58b915a7f11d877ffe4e8e837af6f234d2c/internal/handlers/user/handlers_test.go#L1810).

However, I think we need to add a new test case for this new behaviour, where you'll need to set the value to `false` and where we verify that when this feature is disabled we obtain the expected status code in the response. I think it should be the first test [here](https://github.com/artifacthub/hub/blob/3866c58b915a7f11d877ffe4e8e837af6f234d2c/internal/handlers/user/handlers_test.go#L1094), something like:

```go
t.Run("users sign up is disabled", func(t *testing.T) {
    ...
}
```

**Comment by Skiepp** [comment]
Added 2 tests for the backend part, go test reports all success 👍

**Review by tegioz — APPROVED** [review]
Thanks @Skiepp, looks good to me 👍

**Pull request merged**